src/main/java/ch/vorburger/exec/AtomicExecuteResultHandler.java

```diff
@@ -41,72 +43,68 @@ public class AtomicExecuteResultHandler implements ExecuteResultHandler {
     private static final int SLEEP_TIME_MS = 50;
 
-    private static class Holder {
-        /** The exit value of the finished process */
-        private final Integer exitValue;
+    private final CompletableFuture<Integer> holder = new CompletableFuture<>();
 
-        /** Any offending exception */
-        private final ExecuteException exception;
-
-        Holder(int exitValue) {
-            this.exitValue = exitValue;
-            this.exception = null;
-        }
-
-        Holder(ExecuteException e) {
-            this.exception = e;
-            this.exitValue = null;
-        }
-
-        @Override
-        public String toString() {
-            return "exitValue=" + exitValue + ", exception=" + exception;
-        }
+    private void logOnAlreadySet(String methodName, ExecuteException e) {
+	String errorString = methodName + "  will throw IllegalStateException, already set: " + holder;
+	LOG.error(errorString);
+	throw new IllegalStateException(errorString, e);
     }
 
-    private final AtomicReference<Holder> holder = new AtomicReference<>();
-
     @Override
     public void onProcessComplete(int exitValue) {
-        Holder witness = holder.compareAndExchange(null, new Holder(exitValue));
-        if (witness != null) {
-            LOG.error("onProcessComplete({}) will throw IllegalStateException, already set: {}", exitValue, witness);
-            throw new IllegalStateException(
-                    "Ignoring exit value " + exitValue + " because result already set: " + witness);
-        }
+        if (!holder.complete(exitValue)) {
+	    logOnAlreadySet("onProcessComplete(" + exitValue + ")", null);
+	}
     }
 
     @Override
     public void onProcessFailed(ExecuteException e) {
-        Holder witness = holder.compareAndExchange(null, new Holder(e));
-        if (witness != null) {
-            LOG.error("onProcessFailed({}) will throw IllegalStateException, already set: {}", e, witness);
-            throw new IllegalStateException("Ignoring Exception, because result already set: " + witness, e);
-        }
+	if (!holder.completeExceptionally(e)) {
+	    logOnAlreadySet("onProcessFailed(" + e + ")", e);
+	}
     }
 
     public Optional<Integer> getExitValue() {
-        Holder current = holder.get();
-        return current != null ? Optional.ofNullable(current.exitValue) : Optional.empty();
+	try {
+	    return Optional.ofNullable(holder.getNow(null));
+	} catch (Exception e) {
+	    return Optional.empty();
+	}
     }
 
     public Optional<Exception> getException() {
-        Holder current = holder.get();
-        return current != null ? Optional.ofNullable(current.exception) : Optional.empty();
+	try {
+	    holder.getNow(null);
+	    return Optional.empty();
+	} catch (CompletionException e) {
+	    Throwable inner = e.getCause();
+	    if (inner instanceof ExecuteException) {
+		return Optional.of((ExecuteException) inner);
+	    } else {
+		return Optional.empty();
```

Comment:
This should never happen, and would be a bug if it did - agreed? How about something like:

```suggestion
		throw new IllegalStateException("BUG");
```

Reply:
actually @mosesn let me just merge this as-is, and then follow-up with a PR for this...